File: src/main/scala/io/shinto/amaterasu/execution/actions/runners/spark/PySpark/PySparkEntryPoint.scala
//package io.shinto.amaterasu.execution.actions.runners.spark.PySpark
//
//import java.net.ServerSocket
//import java.util
//
//import io.shinto.amaterasu.runtime.{ AmaContext, Environment }
//
//import org.apache.spark.{ SparkConf, SparkContext, SparkEnv }
//import org.apache.spark.api.java.JavaSparkContext
//import org.apache.spark.sql.SQLContext
//import org.apache.spark.api.python
//
//import py4j.GatewayServer
//
//import scala.collection.concurrent.TrieMap
//
///**
//  * Created by roadan on 10/14/16.
//  */
//object PySparkEntryPoint1 {
//
//  private var started = false
//  private var queue: PySparkExecutionQueue = null
//  private var resultQueues: TrieMap[String, ResultQueue] = null
//  private var port: Int = 0
//  private var jsc: JavaSparkContext = null
//  private var sparkEnv: SparkEnv = null
//
//  def getExecutionQueue(): PySparkExecutionQueue = {
//    queue
//  }
//
//  def getResultQueue(actionName: String): ResultQueue = {
//    resultQueues.getOrElseUpdate(actionName, new ResultQueue)
//  }
//
//  def getJavaSparkContext(): ExtendedJavaContext = {
//    SparkEnv.set(sparkEnv)
//    jsc
//  }
//
//  def getSparkConf(): SparkConf = {
//    jsc.getConf
//  }
//
//  private def generatePort(): Int = {
//
//    val socket = new ServerSocket(0)
//    val port = socket.getLocalPort
//
//    socket.close()
//    port
//
//  }
//
//  def getPort(): Int = {
//    port
//  }
//
//  def start(
//    sc: SparkContext,
//    jobName: String,
//    env: Environment,
//    sparkEnv: SparkEnv
//  ) = {
//
//    if (!started) {
//      AmaContext.init(sc, new SQLContext(sc), jobName, env)
//      started = true
//    }
//
//    jsc = new JavaSparkContext(sc)
//    PySparkEntryPoint1.sparkEnv = sparkEnv
//    queue = new PySparkExecutionQueue()
//    resultQueues = new TrieMap[String, ResultQueue]()
//    port = generatePort()
//    val gatewayServer = new GatewayServer(PySparkEntryPoint, port)
//
//    gatewayServer.start()
//  }
//
//  //seems like we have to do this because of the scala/py4j integration
//  implicit class ExtendedJavaContext(jsc: JavaSparkContext) {
//
//    def sc() = {
//      jsc.sc
//    }
//
//    def stop = jsc.stop
//
//  }
//
//}

File: executor/src/main/java/io/shinto/amaterasu/executor/execution/actions/runners/spark/PySpark/PySparkEntryPoint.java
package io.shinto.amaterasu.executor.execution.actions.runners.spark.PySpark;

import io.shinto.amaterasu.executor.runtime.AmaContext;
import io.shinto.amaterasu.common.runtime.Environment;

import org.apache.spark.SparkEnv;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;

import org.apache.spark.sql.SparkSession;
import py4j.GatewayServer;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.concurrent.ConcurrentHashMap;

public class PySparkEntryPoint {

    //private static Boolean started = false;
    private static PySparkExecutionQueue queue = new PySparkExecutionQueue();
    private static ConcurrentHashMap<String, ResultQueue> resultQueues = new ConcurrentHashMap<>();

    private static int port = 0;
    private static SparkSession sparkSession = null;
    private static JavaSparkContext jsc = null;
    private static SQLContext sqlContext = null;
    private static SparkEnv sparkEnv = null;

    public static PySparkExecutionQueue getExecutionQueue() {
        return queue;
    }

    public static ResultQueue getResultQueue(String actionName) {
        resultQueues.putIfAbsent(actionName, new ResultQueue());
        return resultQueues.get(actionName);
    }

    public static JavaSparkContext getJavaSparkContext() {
        SparkEnv.set(sparkEnv);
        return jsc;
    }

    public static String getJobId(){
        return AmaContext.jobId();
    }

    public static Environment getEnv(){
        return AmaContext.env();
    }

    public static SQLContext getSqlContext() {
        SparkEnv.set(sparkEnv);
        return sqlContext;
    }

    public static SparkSession getSparkSession() {
        SparkEnv.set(sparkEnv);
        return sparkSession;
    }

    public static SparkConf getSparkConf() {
        return jsc.getConf();
    }

    private static void generatePort() {

        try {

            ServerSocket socket = new ServerSocket(0);
            port = socket.getLocalPort();

            socket.close();

        } catch (IOException e) {
        }

    }

    public static int getPort() {
        return port;
    }

    public static void start(SparkSession spark,
                             String jobName,
                             Environment env,
                             SparkEnv sparkEnv) {

        AmaContext.init(spark, jobName, env);

        sparkSession = spark;
        jsc = new JavaSparkContext(spark.sparkContext());
        sqlContext = spark.sqlContext();
        PySparkEntryPoint.sparkEnv = sparkEnv;
        generatePort();
        GatewayServer gatewayServer = new GatewayServer(new PySparkEntryPoint(), port);

        gatewayServer.start();
    }
}
